gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java

```diff
@@ -684,6 +743,187 @@ public Boolean call() {
     }
   }
 
+  @Override
+  public Address get(final AddressId addressId, AddressOption... options) {
+    final Map<ComputeRpc.Option, ?> optionsMap = optionMap(options);
+    try {
+      com.google.api.services.compute.model.Address answer =
+          runWithRetries(new Callable<com.google.api.services.compute.model.Address>() {
+            @Override
+            public com.google.api.services.compute.model.Address call() {
+              switch (addressId.type()) {
+                case REGION:
+                  RegionAddressId regionAddressId = (RegionAddressId) addressId;
```

Comment:
I don't think this is a problem as long as we check and throw if the type is not expected.

How would having interface inheritance per  se avoid casting? To avoid casting we could have two overloaded `get` methods that accept only the available implementations. 

For `create` instead we could have two overloaded `addressId` setters in the `AccessInfo` builder that only accept the available implementations. This way when we reach `create` we are sure that the `addressId` is of a type created by us.